pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -305,28 +305,36 @@ func (s *Scheduler) Solve(ctx context.Context, pods []*corev1.Pod) (Results, err
 		if !ok {
 			break
 		}
-
 		// Schedule to existing nodes or create a new node
-		if podErrors[pod] = s.add(ctx, pod); podErrors[pod] == nil {
-			delete(podErrors, pod)
+		p := pod.DeepCopy()
+		for {
+			if podErrors[pod] = s.add(ctx, pod); podErrors[pod] == nil {
+				delete(podErrors, pod)
+				break
+			}
+			if IsReservedOfferingError(podErrors[pod]) {
+				break
+			}
+			if relaxed := s.preferences.Relax(ctx, pod); !relaxed {
+				break
+			}
+			if err := s.topology.Update(ctx, pod); err != nil && !errors.Is(err, context.DeadlineExceeded) {
```

Comment:
It doesn't need to be coupled to this PR, but we don't need to rebuild the topology for every type of preference we relax, right (e.g. taints / node affinity)?

Reply:
Right -- we could be smarter for sure about it -- I'll take a note to add that as a perf improvement